src/charm.py

```diff
@@ -0,0 +1,222 @@
+        """Handler for the `on_install` event."""
+        self.unit.status = MaintenanceStatus("installing Opensearch Dashboards...")
+
+        install = self.workload.install()
+        if not install:
+            self.unit.status = BlockedStatus("unable to install Opensearch Dashboards")
+
+        # don't complete install until passwords set
+        if not self.state.peer_relation:
+            self.unit.status = WaitingStatus("waiting for peer relation")
+            event.defer()
+            return
+
+        if self.unit.is_leader() and not self.state.cluster.internal_user_credentials:
+            for user in CHARM_USERS:
+                self.state.cluster.update({f"{user}-password": self.workload.generate_password()})
+
+    def _on_cluster_relation_changed(self, event: EventBase) -> None:
+        """Generic handler for all 'something changed, update' events across all relations."""
+        # not all methods called
+        if not self.state.peer_relation:
+            return
+
+        # attempt startup of server
+        if not self.state.unit_server.started:
+            self.init_server()
+
+        # don't delay scale-down leader ops by restarting dying unit
+        if getattr(event, "departing_unit", None) == self.unit:
+            return
+
+        if (
+            self.config_manager.config_changed()
+            and self.state.unit_server.started
+            # and self.upgrade_events.idle
+        ):
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
+
+    def _on_secret_changed(self, event: SecretChangedEvent):
+        """Reconfigure services on a secret changed event."""
+        if not event.secret.label:
+            return
+
+        if not self.state.cluster.relation:
+            return
+
+        if event.secret.label == self.state.cluster.data_interface._generate_secret_label(
+            PEER,
+            self.state.cluster.relation.id,
+            None,  # type:ignore noqa
+        ):  # Changes with the soon upcoming new version of DP-libs STILL within this POC
+
+            if (
+                self.config_manager.config_changed()
+                and self.state.unit_server.started
+                # and self.upgrade_events.idle
+            ):
+                logger.info(f"Secret {event.secret.label} changed.")
+
+    def _start(self, event: EventBase) -> None:
+        """Forces a rolling-restart event.
+
+        Necessary for ensuring that `on_start` restarts roll.
+        """
+        # if not self.state.peer_relation or not self.state.stable or not self.upgrade_events.idle:
+        self.unit.status = MaintenanceStatus("Starting...")
+        if not self.state.peer_relation or not self.state.stable:
+            event.defer()
+            return
+
+        # not needed during application init
+        # only needed for scenarios where the LXD goes down (e.g PC shutdown)
+        if not self.workload.alive():
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
```

Comment:
If dashboard is a stateless app, why do you need a restart lock here? Just start the service.
This step delays the entire deployment, potentially for a long time, and brings no benefits to you. Besides troubleshooting gets more complicated.

The only reason why you need to have a rolling ops is at normal operations, where you want to restart the service without outage for new connections (assuming ofc you have a load balancer in front of the dashboards).

Reply:
Yes, this is being addressed already on [this conversation](https://github.com/canonical/opensearch-dashboards-operator/pull/1/files/49e0214c19a08f3e15997098f48737b1a8804ea4#r1536932210).

Leaving both up until done.